### PR TITLE
Translate CoffeeScript Files To JavaScript (ES2015)

### PR DESCRIPTION
**Pull request by joefitzgerald**
This does what it says on the box. Zero functional changes; just translation of CoffeeScript to JavaScript ES2015. Each non-spec file will be converted, tested with unmodified spec files, then the corresponding spec file(s) will be converted, tested, and pushed. When CI passes, the next non-spec file will go through the same process, until all files are converted.

- [x] Enable `standard` linter
- [x] Fix existing lint issues in `symbol-store.js`
- [x] `autocomplete-manager.coffee`
- [x] `autocomplete-manager-async-spec.coffee`
- [x] `autocomplete-manager-autosave-spec.coffee`
- [x] `autocomplete-manager-integration-spec.coffee`
- [x] `autocomplete-manager-undo-spec.coffee`
- [x] `fuzzy-provider.coffee`
- [x] `fuzzy-provider-spec.coffee`
- [x] `main.js`
- [x] `main-spec.coffee`
- [x] `provider-api-legacy-spec.coffee`
- [x] `provider-api-spec.coffee`
- [x] `private-symbols.coffee`
- [x] `provider-manager.coffee`
- [x] `provider-manager-spec.coffee`
- [x] `provider-metadata.coffee`
- [x] `ref-counted-token-list.coffee`
- [x] `ref-counted-token-list-spec.coffee`
- [x] `scope-helpers.coffee`
- [x] `snippet-parser.coffee`
- [x] `spec-helper.coffee`
- [x] `suggestion-list-element.coffee`
- [x] `suggestion-list-element-spec.coffee`
- [x] `suggestion-list.coffee`
- [x] `symbol-provider.coffee`
- [x] `symbol-provider-spec.coffee`
- [x] `symbol-store.js`
- [x] `symbol-store-spec.coffee`
- [x] `type-helpers.coffee`
- [x] `unicode-helpers.coffee`
- [x] Incorporate changes from #779 
- [x] Incorporate changes from #782
- [x] Incorporate changes from #678

### CODE REVIEW

File: lib/symbol-provider.js [hunk 1]
```diff
@@ -0,0 +1,371 @@
...
+    let bufferEditors
+    let buffer = editor.getBuffer()
+    let editorSubscriptions = new CompositeDisposable()
+
+    // TODO: Remove this conditional once atom/ns-use-display-layers reaches stable and editor.onDidTokenize is always available
+    let onDidTokenizeProvider = (editor.onDidTokenize != null) ? editor : editor.displayBuffer
+
+    editorSubscriptions.add(onDidTokenizeProvider.onDidTokenize(() => {
+      return this.buildWordListOnNextTick(editor)
+    }))
```
Comment (thomasjo) [inline lib/symbol-provider.js]: This block can be changed to
```js
editorSubscriptions.add(onDidTokenizeProvider.onDidTokenize(() => this.buildWordListOnNextTick(editor)))
```

File: lib/symbol-provider.js [hunk 2]
```diff
@@ -0,0 +1,371 @@
...
+    editorSubscriptions.add(editor.onDidDestroy(() => {
+      let index = this.getWatchedEditorIndex(editor)
```
Comment (thomasjo) [inline lib/symbol-provider.js]: Change `let` to `const`.

File: lib/symbol-provider.js [hunk 3]
```diff
@@ -0,0 +1,371 @@
...
+    }))
+
+    bufferEditors = this.watchedBuffers.get(buffer)
+    if (bufferEditors) {
+      return bufferEditors.push(editor)
+    } else {
+      let bufferSubscriptions = new CompositeDisposable()
+      if (bufferSupportsStopChanging()) {
+        bufferSubscriptions.add(buffer.onDidStopChanging(({changes}) => {
+          let editors = this.watchedBuffers.get(buffer)
```
Comment (thomasjo) [inline lib/symbol-provider.js]: Change `let` to `const`.

File: lib/symbol-provider.js [hunk 4]
```diff
@@ -0,0 +1,371 @@
...
+          if (editors && editors.length && editors[0]) {
+            return changes.map(({start, oldExtent, newExtent}) =>
+              this.symbolStore.recomputeSymbolsForEditorInBufferRange(editors[0], start, oldExtent, newExtent))
+          }
+        }))
+      } else {
+        bufferSubscriptions.add(buffer.onDidChange(({oldRange, newRange}) => {
+          let editors = this.watchedBuffers.get(buffer)
```
Comment (thomasjo) [inline lib/symbol-provider.js]: Change `let` to `const`.

File: lib/symbol-provider.js [hunk 5]
```diff
@@ -0,0 +1,371 @@
...
+          if (editors && editors.length && editors[0]) {
+            let { start } = oldRange
```
Comment (thomasjo) [inline lib/symbol-provider.js]: Change `let` to `const`.

File: lib/symbol-provider.js [hunk 6]
```diff
@@ -0,0 +1,371 @@
...
+            let oldExtent = oldRange.getExtent()
```
Comment (thomasjo) [inline lib/symbol-provider.js]: Change `let` to `const`.

File: lib/symbol-provider.js [hunk 7]
```diff
@@ -0,0 +1,371 @@
...
+            let newExtent = newRange.getExtent()
```
Comment (thomasjo) [inline lib/symbol-provider.js]: Change `let` to `const`.

File: lib/symbol-provider.js [hunk 8]
```diff
@@ -0,0 +1,371 @@
...
+  isWatchingEditor (editor) {
+    return this.getWatchedEditorIndex(editor) > -1
+  }
+
+  isWatchingBuffer (buffer) {
+    return (this.watchedBuffers.get(buffer) != null)
+  }
+
+  getWatchedEditorIndex (editor) {
+    let editors = this.watchedBuffers.get(editor.getBuffer())
```
Comment (thomasjo) [inline lib/symbol-provider.js]: Change `let` to `const`.

File: lib/symbol-provider.js [hunk 9]
```diff
@@ -0,0 +1,371 @@
...
+    if (!this.scopeDescriptorsEqual(this.configScopeDescriptor, scopeDescriptor)) {
+      this.buildConfig(scopeDescriptor)
+      this.configScopeDescriptor = scopeDescriptor
+      return this.configScopeDescriptor
+    }
+  }
+
+  buildConfig (scopeDescriptor) {
+    this.config = {}
+    let legacyCompletions = this.settingsForScopeDescriptor(scopeDescriptor, 'editor.completions')
```
Comment (thomasjo) [inline lib/symbol-provider.js]: Change `let` to `const`.

File: lib/symbol-provider.js [hunk 10]
```diff
@@ -0,0 +1,371 @@
...
+    let allConfigEntries = this.settingsForScopeDescriptor(scopeDescriptor, 'autocomplete.symbols')
+
+    // Config entries are reverse sorted in order of specificity. We want most
+    // specific to win; this simplifies the loop.
+    allConfigEntries.reverse()
+
+    for (let i = 0; i < legacyCompletions.length; i++) {
+      let {value} = legacyCompletions[i]
```
Comment (thomasjo) [inline lib/symbol-provider.js]: Change `let` to `const`. Also add some whitespace inside the brackets to be consistent with other, similar code?

File: lib/symbol-provider.js [hunk 11]
```diff
@@ -0,0 +1,371 @@
...
+      if (Array.isArray(value) && value.length) { this.addLegacyConfigEntry(value) }
+    }
+
+    let addedConfigEntry = false
+    for (let j = 0; j < allConfigEntries.length; j++) {
+      let {value} = allConfigEntries[j]
```
Comment (thomasjo) [inline lib/symbol-provider.js]: Change `let` to `const`. Same whitespace remark as earlier.

File: lib/symbol-provider.js [hunk 12]
```diff
@@ -0,0 +1,371 @@
...
+
+  addLegacyConfigEntry (suggestions) {
+    suggestions = (suggestions.map((suggestion) => ({text: suggestion, type: 'builtin'})))
+    if (this.config.builtin == null) { this.config.builtin = {suggestions: []} }
+    this.config.builtin.suggestions = this.config.builtin.suggestions.concat(suggestions)
+    return this.config.builtin.suggestions
+  }
+
+  addConfigEntry (config) {
+    for (let type in config) {
```
Comment (thomasjo) [inline lib/symbol-provider.js]: Change `let` to `const`.

File: lib/symbol-provider.js [hunk 13]
```diff
@@ -0,0 +1,371 @@
...
+      let options = config[type]
```
Comment (thomasjo) [inline lib/symbol-provider.js]: Change `let` to `const`.

File: lib/symbol-provider.js [hunk 14]
```diff
@@ -0,0 +1,371 @@
...
+      if (this.config[type] == null) { this.config[type] = {} }
+      if (options.selector != null) { this.config[type].selectors = Selector.create(options.selector) }
+      this.config[type].typePriority = options.typePriority != null ? options.typePriority : 1
+      this.config[type].wordRegex = this.wordRegex
+
+      let suggestions = this.sanitizeSuggestionsFromConfig(options.suggestions, type)
```
Comment (thomasjo) [inline lib/symbol-provider.js]: Change `let` to `const`.

File: lib/symbol-provider.js [hunk 15]
```diff
@@ -0,0 +1,371 @@
...
+    if (!options.prefix) {
+      return
+    }
+
+    if (options.prefix.trim().length < this.minimumWordLength) {
+      return
+    }
+
+    this.buildConfigIfScopeChanged(options)
+    let editor = options.editor
```
Comment (thomasjo) [inline lib/symbol-provider.js]: Change `let` to `const`.

File: lib/symbol-provider.js [hunk 16]
```diff
@@ -0,0 +1,371 @@
...
+    let bufferPosition = options.bufferPosition
```
Comment (thomasjo) [inline lib/symbol-provider.js]: Change `let` to `const`.

File: lib/symbol-provider.js [hunk 17]
```diff
@@ -0,0 +1,371 @@
...
+    let prefix = options.prefix
```
Comment (thomasjo) [inline lib/symbol-provider.js]: Change `let` to `const`.

File: lib/symbol-provider.js [hunk 18]
```diff
@@ -0,0 +1,371 @@
...
+
+    let numberOfWordsMatchingPrefix = 1
+    let wordUnderCursor = this.wordAtBufferPosition(editor, bufferPosition)
```
Comment (thomasjo) [inline lib/symbol-provider.js]: Change `let` to `const`.

File: lib/symbol-provider.js [hunk 19]
```diff
@@ -0,0 +1,371 @@
...
+    let iterable = editor.getCursors()
```
Comment (thomasjo) [inline lib/symbol-provider.js]: Change `let` to `const`.

File: lib/symbol-provider.js [hunk 20]
```diff
@@ -0,0 +1,371 @@
...
+    for (let i = 0; i < iterable.length; i++) {
+      let cursor = iterable[i]
+      if (cursor === editor.getLastCursor()) { continue }
+      let word = this.wordAtBufferPosition(editor, cursor.getBufferPosition())
+      if (word === wordUnderCursor) { numberOfWordsMatchingPrefix += 1 }
+    }
+
+    let buffers = this.includeCompletionsFromAllBuffers ? null : [this.editor.getBuffer()]
```
Comment (thomasjo) [inline lib/symbol-provider.js]: Change `let` to `const`.

File: lib/symbol-provider.js [hunk 21]
```diff
@@ -0,0 +1,371 @@
...
+    let symbolList = this.symbolStore.symbolsForConfig(
+      this.config, buffers,
+      prefix, wordUnderCursor,
+      bufferPosition.row,
+      numberOfWordsMatchingPrefix
+    )
```
Comment (thomasjo) [inline lib/symbol-provider.js]: Is there a reason for this somewhat unorthodox formatting? This should be broken on every comma, right? Alternatively break once, and leave all parameters on a single line.

File: lib/symbol-provider.js [hunk 22]
```diff
@@ -0,0 +1,371 @@
...
+
+    symbolList.sort((a, b) => (b.score * b.localityScore) - (a.score * a.localityScore))
+    return symbolList.slice(0, 20).map(a => a.symbol)
+  }
+
+  wordAtBufferPosition (editor, bufferPosition) {
+    let lineToPosition = editor.getTextInRange([[bufferPosition.row, 0], bufferPosition])
```
Comment (thomasjo) [inline lib/symbol-provider.js]: Change `let` to `const`.

File: lib/symbol-provider.js [hunk 23]
```diff
@@ -0,0 +1,371 @@
...
+    let prefix = lineToPosition.match(this.endOfLineWordRegex)
+    if (prefix) {
+      prefix = prefix[0]
+    } else {
+      prefix = ''
+    }
+
+    let lineFromPosition = editor.getTextInRange([bufferPosition, [bufferPosition.row, Infinity]])
```
Comment (thomasjo) [inline lib/symbol-provider.js]: Change `let` to `const`.

File: lib/symbol-provider.js [hunk 24]
```diff
@@ -0,0 +1,371 @@
...
+  Section: Word List Building
+  */
+
+  buildWordListOnNextTick (editor) {
+    return _.defer(() => {
+      if (!editor || !editor.isAlive()) {
+        return
+      }
+
+      let start = {row: 0, column: 0}
```
Comment (thomasjo) [inline lib/symbol-provider.js]: Change `let` to `const`.

File: lib/symbol-provider.js [hunk 25]
```diff
@@ -0,0 +1,371 @@
...
+      let oldExtent = {row: 0, column: 0}
```
Comment (thomasjo) [inline lib/symbol-provider.js]: Change `let` to `const`.

File: lib/symbol-provider.js [hunk 26]
```diff
@@ -0,0 +1,371 @@
...
+      let newExtent = editor.getBuffer().getRange().getExtent()
```
Comment (thomasjo) [inline lib/symbol-provider.js]: Change `let` to `const`.

File: lib/symbol-provider.js [hunk 27]
```diff
@@ -0,0 +1,371 @@
...
+    }
+    )
+  }
+
+  // FIXME: this should go in the core ScopeDescriptor class
+  scopeDescriptorsEqual (a, b) {
+    if (a === b) { return true }
+    if ((a == null) || (b == null)) { return false }
+
+    let arrayA = a.getScopesArray()
```
Comment (thomasjo) [inline lib/symbol-provider.js]: Change `let` to `const`.

File: lib/symbol-provider.js [hunk 28]
```diff
@@ -0,0 +1,371 @@
...
+    let arrayB = b.getScopesArray()
```
Comment (thomasjo) [inline lib/symbol-provider.js]: Change `let` to `const`.

File: lib/symbol-store.js [hunk 29]
```diff
@@ -102,11 +102,11 @@ export default class SymbolStore {
...
     let tokenizedBuffer = editor.tokenizedBuffer ? editor.tokenizedBuffer : editor.displayBuffer.tokenizedBuffer
 
-    for (var bufferRow = start.row; bufferRow <= newEnd; bufferRow++) {
+    for (let bufferRow = start.row; bufferRow <= newEnd; bufferRow++) {
       let tokenizedLine = tokenizedBuffer.tokenizedLineForRow(bufferRow)
       if (tokenizedLine == null) continue
 
-      let symbolsByLetter = new Map
+      let symbolsByLetter = new Map()
       let tokenIterator = tokenizedLine.getTokenIterator()
```
Comment (thomasjo) [inline lib/symbol-store.js]: Change `let` to `const`.

File: lib/symbol-store.js [hunk 30]
```diff
@@ -102,11 +102,11 @@ export default class SymbolStore {
...
       while (tokenIterator.next()) {
         let wordsWithinToken = tokenIterator.getText().match(this.wordRegex) || EMPTY_ARRAY
```
Comment (thomasjo) [inline lib/symbol-store.js]: Change `let` to `const`.

File: lib/type-helpers.js [hunk 31]
```diff
@@ -0,0 +1,13 @@
+'use babel'
+
+let isFunction = value => isType(value, 'function')
```
Comment (thomasjo) [inline lib/type-helpers.js]: Change `let` to `const`.

File: lib/type-helpers.js [hunk 32]
```diff
@@ -0,0 +1,13 @@
...
+
+let isString = value => isType(value, 'string')
```
Comment (thomasjo) [inline lib/type-helpers.js]: Change `let` to `const`.

File: lib/type-helpers.js [hunk 33]
```diff
@@ -0,0 +1,13 @@
...
+
+var isType = function (value, typeName) {
```
Comment (thomasjo) [inline lib/type-helpers.js]: Just make this a "real" function instead? `function isType (value, typeName) { `
I guess the same goes for the functions above, but at least here it avoids re-ordering and means we can drop the `var` specifier.

File: lib/type-helpers.js [hunk 34]
```diff
@@ -0,0 +1,13 @@
...
+  let t = typeof value
```
Comment (thomasjo) [inline lib/type-helpers.js]: Change `let` to `const`.

**Comment by as-cii** [comment]
@joefitzgerald just a quick heads-up: as we discussed I have merged #782, so there's going to be some merge conflict. Please, ping me if you need any help back porting the changes I made in CoffeeScript. Thanks!

**Comment by joefitzgerald** [comment]
@as-cii I incorporated the changes from #782 in 5be853f0d6af4275f8cd41b292073b8da6988ac2; it's worth noting that all the builds for the intermediate commits were green prior to rebasing. I am inclined to merge #779 now to minimize further merge conflicts. You cool with that?

**Comment by joefitzgerald** [comment]
@thomasjo I addressed your review comments; please review the whole PR now, as it is complete. Obviously we could improve the code further, but I don't want to focus on picking up every possible improvement, and perhaps address additional work (if significant) in subsequent PRs.

**Comment by as-cii** [comment]
Thank you so much @joefitzgerald!!! ❤️ ❤️ ❤️

**Pull request merged**